BEPs/BEP-334.md

```diff
@@ -0,0 +1,100 @@
+<pre>
+  BEP: 334
+  Title: Greenfield CrossChain Permission Module
+  Status: Draft
+  Type: Standards
+  Created: 2023-12-06
+</pre>
+
+# BEP-334: Greenfield CrossChain Permission Module
+
+<!-- @import "[TOC]" {cmd="toc" depthFrom=1 depthTo=6 orderedList=false} -->
+
+<!-- code_chunk_output -->
+
+- [BEP-334: Greenfield CrossChain Permission Module](#bep-334-greenfield-crosschain-permission-module)
+    - [1. Summary](#1-summary)
+    - [2. Abstract](#2-abstract)
+    - [3. Motivation](#3-motivation)
+    - [4. Specification](#4-specification)
+      - [4.1 Definitions](#41-definitions)
+      - [4.2 Interface](#42-interface)
+        - [4.2.1 Create Policy](#421-create-policy)
+        - [4.2.2 Delete Policy](#422-delete-policy)
+    - [5. License](#5-license)
+
+<!-- /code_chunk_output -->
+
+
+
+## 1. Summary
+Currently, Greenfield supports creating buckets and groups from the smart contracts deployed on BSC/opBNB. This means that the owner of a bucket can be a contract address. However, contract addresses cannot upload objects or put policies under the bucket, resulting in empty buckets owned by smart contracts.
+To address this issue, we are going to introduce the cross-chain permission module. With this BEP, smart contracts can grant EOA addresses the ability to upload objects under their buckets, and even support additional functions.
+
+## 2. Abstract
+Greenfield CrossChain Permission Module has several significant differences and improvements compared to current implementations:
+- BSC/opBNB account is allowed to put policy for Greenfield resources.
+- BSC/opBNB account is allowed to disable policy for Greenfield resources. 
+
+## 3. Motivation
+In the current framework, users could only change resource permission on Greenfield. The proposal allows Greenfield users to change resource permission directly on the BSC/opBNB network.
+
+## 4. Specification
+### 4.1 Definitions
+- PermissionHub Contract: A new middle-layer contract to request permission changes from BSC to Greenfield
+- 
+### 4.2 Interface
+#### 4.2.1 Create Policy
+The interface to create policy is as follows:
+**function createPolicy(bytes calldata _data, ExtraData memory _extraData) external;**
+`_data` is the RLP encoded bytes of the struct MsgPutPolicy as follows:
+```golang
+type MsgPutPolicy struct {
```

Comment:
I will suggest use proto rather than go code here.

Reply:
```

message Policy {
  string id = 1 [
    (cosmos_proto.scalar) = "cosmos.Uint",
    (gogoproto.customtype) = "Uint",
    (gogoproto.nullable) = false
  ];
  Principal principal = 2;
  resource.ResourceType resource_type = 3;
  string resource_id = 4 [
    (cosmos_proto.scalar) = "cosmos.Uint",
    (gogoproto.customtype) = "Uint",
    (gogoproto.nullable) = false
  ];
  repeated Statement statements = 5;
  google.protobuf.Timestamp expiration_time = 6 [
    (gogoproto.stdtime) = true,
    (gogoproto.nullable) = true
  ];
}
```

proto has a lot of native type dependencies that are redundant to explain